jvm/src/main/java/com/microsoft/bond/protocol/CompactBinaryProtocolWriter.java

```diff
@@ -0,0 +1,383 @@
+package com.microsoft.bond.protocol;
+
+import com.microsoft.bond.BondDataType;
+import com.microsoft.bond.Metadata;
+import com.microsoft.bond.ProtocolType;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.Queue;
+import java.util.Stack;
+
+// TODO: add Compact Binary protocol format comment (either here and/or in the deserializer and/or in package.html)
+
+/**
+ * Implements Compact Binary protocol (serialization).
+ */
+public final class CompactBinaryProtocolWriter implements TwoPassProtocolWriter {
+
+    private static final short MAGIC = (short)ProtocolType.COMPACT_PROTOCOL.value;
+
+    private final BinaryStreamWriter writer;
+    private final short version;
+    private final FirstPassCounter firstPassCounter;
+
+    private int currentStructIndex;
+
+    public CompactBinaryProtocolWriter(OutputStream outputStream, short version) {
+        this.writer = new BinaryStreamWriter(outputStream);
+        this.version = version;
+        this.firstPassCounter = (version == 2) ? new FirstPassCounter() : null;
+        this.currentStructIndex = 0;
+    }
+
+    @Override
+    public void writeVersion() throws IOException {
+        this.writer.writeInt16(MAGIC);
+        this.writer.writeInt16(this.version);
+    }
+
+    @Override
+    public void writeStructBegin(Metadata metadata) throws IOException {
+        if (this.version == 2) {
+            int structLength = this.firstPassCounter.getResultLength(this.currentStructIndex);
+            this.writer.writeVarUInt32(structLength);
+        }
+
+        ++this.currentStructIndex;
+    }
+
+    @Override
+    public void writeBaseBegin(Metadata metadata) throws IOException {}
+
+    @Override
+    public void writeStructEnd() throws IOException {
+        this.writer.writeInt8((byte)BondDataType.BT_STOP.value);
+
+        --this.currentStructIndex;
+    }
+
+    @Override
+    public void writeBaseEnd() throws IOException {
+        this.writer.writeInt8((byte)BondDataType.BT_STOP_BASE.value);
+    }
+
+    @Override
+    public void writeFieldBegin(BondDataType type, int id, Metadata metadata) throws IOException {
+        if (id <= 5) {
+            this.writer.writeInt8((byte)(type.value | (id << 5)));
+        }
+        else if (id <= 0xFF) {
```

Comment:
~~ This is an interesting hybrid of MS and non-MS brace style. :) `} else {`, please.